Add ContainerRowSerde::compareWithNulls API

duanmeng opened the pull request:
Presto aggregate functions that compare values of complex types need to detect
cases when complex type values contain nulls. The new
ContainerRowSerde::compareWithNulls API allows to signal the case when complex
type value contains a null.

Part of #6314

mbasmanova commented on velox/exec/ContainerRowSerde.cpp:
```diff
@@ -373,37 +372,56 @@ int compareStringAsc(
 }
 
+std::optional<std::optional<int32_t>>
+compareNulls(bool leftNull, bool rightNull, CompareFlags flags) {
```
Is this a copy of BaseVector::compareNulls? If so, maybe change BaseVector::compareNulls to be a static public method and re-use.

duanmeng replied:
Yes, we could reuse `BaseVector::compareNulls`. Updated.

mbasmanova commented on velox/exec/ContainerRowSerde.cpp:
```diff
@@ -373,37 +372,56 @@ int compareStringAsc(
   return flags.ascending ? result : result * -1;
 }
 
+std::optional<std::optional<int32_t>>
```
Double optional is a bit hard to reason about. Is this necessary?

duanmeng replied:
Reuse `BaseVector::compareNulls` now.

mbasmanova commented on velox/exec/tests/ContainerRowSerdeTest.cpp:
```diff
@@ -144,5 +145,91 @@ TEST_F(ContainerRowSerdeTest, nested) {
+  DecodedVector decodedVector;
+  decodedVector.decode(*rowVector, rows);
+
+  static const CompareFlags kCompareFlags{
```
It is confusing that we serializes a value of type ROW(INTEGER, VARCHAR), but compare it to a value of type ROW(INTEGER).

duanmeng replied:
Fixed

mbasmanova commented on velox/exec/tests/ContainerRowSerdeTest.cpp:
```diff
@@ -144,5 +145,91 @@ TEST_F(ContainerRowSerdeTest, nested) {
+      true, // nullsFirst
+      true, // ascending
+      false, // equalsOnly
+      CompareFlags::NullHandlingMode::StopAtNull};
```
I think we need more test coverage, i.e. verify all values of NullHandlingMode and combinations with different values of equalsOnly.

duanmeng replied:
Sure, I've refactored the ut, and imporved the coverage.

mbasmanova commented on velox/exec/tests/ContainerRowSerdeTest.cpp:
```diff
@@ -144,5 +145,91 @@ TEST_F(ContainerRowSerdeTest, nested) {
+  ByteStream out;
+  HashStringAllocator::prepareRead(position.header, out);
+
+  auto arrayVector = makeNullableArrayVector<int64_t>({
```
Would you also add tests for maps?

duanmeng replied:
Added

mbasmanova reviewed (commented):
@laithsakka Laith, would you help review this PR?

kagamiori commented on velox/exec/ContainerRowSerde.h:
```diff
@@ -44,6 +44,15 @@ class ContainerRowSerde {
 
+  /// Return std::nullopt if encountered in NullHandlingMode::StopAtNull mode.
+  /// Support NullHandlingMode::NoStop mode as well and return
+  /// std::optional<int32_t>(0) if null encountered.
```
I may have missed this, but where do we return std::nullopt in the leaf cases?

duanmeng replied:
I'm not sure I understand your question about ** leaf cases**. 

It will return `std::nullopt`, say when an element of an ArrayVector is null during in StopAtAtNull mode. `BaseVector::compareNulls` returns `std::nullopt` in StopAtAtNull mode when a null is encountered.
```C++
std::optional<int32_t> compareArrays(
    ...
    if (leftNull || rightNull) {
      auto result = BaseVector::compareNulls(leftNull, rightNull, flags);
      if (result && *result == 0) {
        continue;
      }
      return result;
    }
```

duanmeng replied:
> /// Return std::nullopt if encountered in NullHandlingMode::StopAtNull mode.
  /// Support NullHandlingMode::NoStop mode as well and return
  /// std::optional<int32_t>(0) if null encountered.

This is not clear and may be wrong. I will update it as follows,
"Return std::nullopt if encountered in NullHandlingMode::StopAtNull mode. And return std::optional<int32_t>(0) if both sides are nulls in NullHandlingMode::NoStop mode."

kagamiori commented on velox/exec/ContainerRowSerde.cpp:
```diff
@@ -522,33 +514,29 @@ int compare<TypeKind::ARRAY>(
   auto rightIndices = map->sortedKeyIndices(wrappedIndex);
   auto result = compareArrayIndices(left, *map->mapKeys(), rightIndices, flags);
-  if (result) {
+  if (result.has_value() && result.value() != 0) {
```
nit: `if (result && *result != 0)` for consistency with other places?

duanmeng replied:
Thanks for your remind. I've fixed it using the following code, and improve the ut of MapType.
```C++
if (result && *result == 0) {
    return compareArrayIndices(left, *map->mapValues(), rightIndices, flags);
  }
  return result;
```

kagamiori reviewed (commented):
Hi @duanmeng, Thank you for adding this support to fix min() and max() functions! I left two questions.

duanmeng commented on velox/exec/ContainerRowSerde.h:
```diff
@@ -44,6 +44,12 @@ class ContainerRowSerde {
       const Type* type,
       CompareFlags flags);
 
+  static std::optional<int32_t> compareWithNulls(
```
Fixed

duanmeng commented on velox/exec/tests/ContainerRowSerdeTest.cpp:
```diff
@@ -144,5 +145,91 @@ TEST_F(ContainerRowSerdeTest, nested) {
+  std::optional<std::optional<int>> x =
+      std::make_optional<std::optional<int>>(std::nullopt);
+  auto data = makeRowVector({
+      makeNullableFlatVector<int32_t>({
```
Fixed

duanmeng commented on velox/exec/tests/ContainerRowSerdeTest.cpp:
```diff
@@ -144,5 +145,91 @@ TEST_F(ContainerRowSerdeTest, nested) {
 }
 
+TEST_F(ContainerRowSerdeTest, compareNullsInRowVector) {
+  std::optional<std::optional<int>> x =
```
Removed

duanmeng commented:
@mbasmanova Hi Masha, I've fixed all the comments, refactored the UT codes, and improved the coverage according to your guidance. Could you help to take a look?

mbasmanova commented on velox/exec/ContainerRowSerde.cpp:
```diff
@@ -415,31 +408,30 @@ int compare<TypeKind::ROW>(
+
+    if (leftNull || rightNull) {
+      auto result = BaseVector::compareNulls(leftNull, rightNull, flags);
+      if (result && *result == 0) {
```
I find this a bit hard to read. Consider being more explicit:

```
if (result.has_value() && result.value() == 0)
```

duanmeng replied:
Fixed

mbasmanova commented on velox/exec/ContainerRowSerde.cpp:
```diff
@@ -415,31 +408,30 @@ int compare<TypeKind::ROW>(
     auto result = compareSwitch(left, *child, wrappedIndex, flags);
-    if (result) {
-      return result;
+    if (result && *result == 0) {
```
same

duanmeng replied:
Fixed

mbasmanova commented on velox/exec/ContainerRowSerde.h:
```diff
@@ -44,6 +44,15 @@ class ContainerRowSerde {
       const Type* type,
       CompareFlags flags);
 
+  /// Return std::nullopt if encountered in NullHandlingMode::StopAtNull mode.
```
I find this comment a bit cryptic. Maybe, use the comment from BaseVector::compare as a base.

```
 /// Returns < 0 if 'left' is less than 'right' at 'index', 0 if
 /// equal and > 0 otherwise. If flags.nullHandlingMode is StopAtNull, 
 /// returns std::nullopt if either 'left' or 'right' value is null or contains a null.
```

Also, let's document existing 'compare' methods to clarify that these support only NoStop for flags.nullHandlingMode, consider NULL equal to NULL.

duanmeng replied:
Added

mbasmanova commented on velox/exec/tests/ContainerRowSerdeTest.cpp:
```diff
@@ -35,6 +35,39 @@ class ContainerRowSerdeTest : public testing::Test,
     return position;
   }
 
+  std::vector<std::shared_ptr<ByteStream>> serialize(
```
This method can be simplified to just serialize the data and return a list of positions. No need to create and return ByteStreams. These can be created on the fly from the positions in testCompareWithNulls. I think this would also remove the need for rewindByteStreams method.

duanmeng replied:
Thanks for your suggestion, It is much cleaner now.

mbasmanova commented on velox/exec/tests/ContainerRowSerdeTest.cpp:
```diff
@@ -62,6 +95,31 @@ class ContainerRowSerdeTest : public testing::Test,
+  void testCompareWithNulls(
+      const DecodedVector& decodedVector,
+      const std::vector<HashStringAllocator::Position>& positions,
+      const std::vector<std::shared_ptr<ByteStream>>& byteStreams,
```
Per comment above, let's remove byteStreams parameter and create ByteStreeams from positions on the fly.

duanmeng replied:
Fixed

mbasmanova commented on velox/exec/tests/ContainerRowSerdeTest.cpp:
```diff
@@ -62,6 +95,31 @@ class ContainerRowSerdeTest : public testing::Test,
+          expected.at(i),
+          ContainerRowSerde::compareWithNulls(
+              *byteStreams.at(i), decodedVector, i, compareFlags)
+              .value());
```
This assumes that compareWithNulls cannot return std::nullopt? Is this a valid assumption? Don't we want to allow testing cases where compareWithNulls returns std::nullopt?

duanmeng replied:
Fixed

mbasmanova commented on velox/exec/tests/ContainerRowSerdeTest.cpp:
```diff
@@ -62,6 +95,31 @@ class ContainerRowSerdeTest : public testing::Test,
+    }
+
+    auto res = ContainerRowSerde::compareWithNulls(
+        *byteStreams.at(3), decodedVector, 3, compareFlags);
```
Hard-coding index '3' here makes this method hard to use. Let's change expected to std::vector<std::optional<int32_t>> and remove expectedOptionVal.

duanmeng replied:
Fixed

mbasmanova commented on velox/exec/tests/ContainerRowSerdeTest.cpp:
```diff
@@ -144,5 +202,142 @@ TEST_F(ContainerRowSerdeTest, nested) {
+      byteStreams,
+      {0, 1, -1},
+      std::nullopt,
+      compareFlags);
```
Since we only test different values of equalsOnly and nullHandlingMode, the test may be more readable if we change testCompareWithNulls to take these 2 values instead of CompareFlags.

duanmeng replied:
Fixed

mbasmanova reviewed (commented):
@duanmeng Thank you for iterating on this PR. This is shaping up nicely. Some comments below.

mbasmanova commented on velox/exec/ContainerRowSerde.h:
```diff
@@ -32,18 +32,35 @@ class ContainerRowSerde {
+  /// Returns < 0 if 'left' is less than 'right' at 'index', 0 if
+  /// equal and > 0 otherwise. If flags.nullHandlingMode is StopAtNull,
+  /// returns std::nullopt if either 'left' or 'right' value is null or contains
+  /// a null. If flags.nullHandlingMode is NoStop, returns
```
> If flags.nullHandlingMode is NoStop, returns
> std::optional<int32_t>(0) if both sides are nulls

Let's remove this sentence. It is missing in comments for other versions of this method.

Alternatively, let's say that "If flags.nullHandlingMode is NoStop then NULL is considered equal to NULL."

mbasmanova commented on velox/exec/tests/ContainerRowSerdeTest.cpp:
```diff
@@ -35,6 +35,26 @@ class ContainerRowSerdeTest : public testing::Test,
     return position;
   }
 
+  std::vector<HashStringAllocator::Position> serializeWithPositions(
```
Is it necessary to add WithPositions to the name of this method? Can it be just 'serialize'?

duanmeng replied:
There is another method named serialize with the same input parameters.

mbasmanova replied:
I see. An existing serialize() writes all rows together and returns a position at the start of this combined write. This one writes each row individually and returns positions for individual rows.

Perhaps, we could remove the existing method and rename this one to 'serialize' + add a boolean indicating whether to serialize all rows together and return a single 'position' or serialize each row individually and returns per-row positions.

Either way, it would be helpful to add comments to clarify how these 2 methods differ from each other.

Not a big deal though.

duanmeng replied:
I've added comments to clarify the difference.

mbasmanova commented on velox/exec/tests/ContainerRowSerdeTest.cpp:
```diff
@@ -35,6 +35,26 @@ class ContainerRowSerdeTest : public testing::Test,
+      ContainerRowSerde::serialize(*data, i, out);
+      allocator_.finishWrite(out, 0);
+
+      auto cur = std::make_shared<ByteStream>();
```
The logic using 'cur' is not needed. Let's remove.

duanmeng replied:
Fixed

mbasmanova commented on velox/exec/tests/ContainerRowSerdeTest.cpp:
```diff
@@ -62,6 +82,28 @@ class ContainerRowSerdeTest : public testing::Test,
+        mode};
+
+    for (auto i = 0; i < expected.size(); ++i) {
+      ByteStream buf;
```
buf -> stream

duanmeng replied:
Fixed

mbasmanova commented on velox/exec/tests/ContainerRowSerdeTest.cpp:
```diff
@@ -144,5 +186,126 @@ TEST_F(ContainerRowSerdeTest, nested) {
+      {1, 2, 3, 4},
+  });
+
+  std::vector<HashStringAllocator::Position> positions =
```
auto

duanmeng replied:
Fixed

mbasmanova commented on velox/exec/tests/ContainerRowSerdeTest.cpp:
```diff
@@ -144,5 +186,126 @@ TEST_F(ContainerRowSerdeTest, nested) {
+TEST_F(ContainerRowSerdeTest, compareNullsInArrayVector) {
+  auto data = makeArrayVector<int64_t>({
+      {1, 2},
+      {1, 5},
```
Maybe add some arrays with null elements?

duanmeng replied:
Added

mbasmanova commented on velox/exec/tests/ContainerRowSerdeTest.cpp:
```diff
@@ -144,5 +186,126 @@ TEST_F(ContainerRowSerdeTest, nested) {
+      serializeWithPositions(data);
+
+  auto someNulls = makeNullableFlatVector<int32_t>({
+      {1},
```
not need for {} around numbers

duanmeng replied:
Fixed

mbasmanova commented on velox/exec/tests/ContainerRowSerdeTest.cpp:
```diff
@@ -144,5 +186,126 @@ TEST_F(ContainerRowSerdeTest, nested) {
+  auto rowVector = makeRowVector({someNulls});
+  SelectivityVector rows(rowVector->size());
+  DecodedVector decodedVector;
+  decodedVector.decode(*rowVector, rows);
```
You can remove 'rows' parameter and then remove 'rows' variable.

DecodedVector decodedVector(*decodedVector);

duanmeng replied:
Fixed

mbasmanova reviewed (commented):
@duanmeng Look great % a few remaining nits.

mbasmanova commented on velox/exec/tests/ContainerRowSerdeTest.cpp:
```diff
@@ -144,5 +183,121 @@ TEST_F(ContainerRowSerdeTest, nested) {
+      {1, 2, std::nullopt, 4},
+      {1, 5},
+  });
+  DecodedVector decodedVector;
```
You can combine these 2 lines into one:

```
DecodedVector decodedVector(*arrayVector);
```

Ditto other places.

duanmeng replied:
Fixed
